File: test/core/data/repositories/supabase_auth_repository_test.dart
import 'dart:convert';

import 'package:daypilot/core/data/models/auth_exceptions.dart';
import 'package:daypilot/core/data/repositories/supabase_auth_repository.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:supabase_flutter/supabase_flutter.dart';

import '../../../support/fake_supabase_http.dart';

/// Builds a GoTrue-shaped success body: a session (so `AuthResponse.fromJson`
/// resolves both `.session` and `.user`) wrapping the given user id.
Map<String, dynamic> _sessionBody(String userId, {Map<String, dynamic>? userMetadata}) {
  final exp = DateTime.now().add(const Duration(hours: 1)).millisecondsSinceEpoch ~/ 1000;
  final header = base64Url.encode(utf8.encode('{"alg":"HS256","typ":"JWT"}'));
  final payload = base64Url.encode(utf8.encode('{"exp":$exp,"sub":"$userId"}'));
  return {
    'access_token': '$header.$payload.signature',
    'token_type': 'bearer',
    'expires_in': 3600,
    'refresh_token': 'refresh-$userId',
    'user': {
      'id': userId,
      'app_metadata': <String, dynamic>{},
      'user_metadata': userMetadata ?? <String, dynamic>{},
      'aud': 'authenticated',
      'created_at': '2024-01-01T00:00:00Z',
    },
  };
}

void main() {
  late FakeSupabaseHttp http;
  late SupabaseAuthRepository repo;

  setUp(() {
    http = FakeSupabaseHttp();
    repo = SupabaseAuthRepository(http.client);
  });

  tearDown(() => http.dispose());

  group('currentUser', () {
    test('with no signed-in user returns null and makes no request', () async {
      final user = await repo.currentUser();

      expect(user, isNull);
      expect(http.requests, isEmpty);
    });

    test('with an existing profile row skips the insert', () async {
      await http.signIn('user-1');
      http.enqueue(jsonRows([
        {'id': 'user-1'},
      ]));
      http.enqueue(jsonRow({
        'id': 'user-1',
        'email': 'ana@daypilot.test',
        'name': 'Ana',
        'username': 'ana',
        'photo_url': null,
        'region': null,
        'level': 3,
      }));

      final user = await repo.currentUser();

      expect(user!.username, 'ana');
      expect(http.requests, hasLength(2));
    });

    test('with no profile row yet creates one from user metadata', () async {
      await http.signIn('user-1');
      http.enqueue(emptyRows());
      http.enqueue(noContent());
      http.enqueue(jsonRow({
        'id': 'user-1',
        'email': 'ana@daypilot.test',
        'name': 'Ana',
        'username': 'ana',
        'photo_url': null,
        'region': null,
        'level': 1,
      }));

      final user = await repo.currentUser();

      expect(user!.level, 1);
      expect(http.requests, hasLength(3));
      expect(http.requests[1].method, 'POST');
      expect(http.requests[1].url.path, '/rest/v1/users');
    });
  });

  group('login', () {
    test('signs in and fetches the existing profile', () async {
      http.enqueue(jsonRow(_sessionBody('user-1')));
      http.enqueue(jsonRows([
        {'id': 'user-1'},
      ]));
      http.enqueue(jsonRow({
        'id': 'user-1',
        'email': 'ana@daypilot.test',
        'name': 'Ana',
        'username': 'ana',
        'photo_url': null,
        'region': null,
        'level': 2,
      }));

      final user = await repo.login(email: 'ana@daypilot.test', password: 'password123');

      expect(user.username, 'ana');
      expect(http.requests[0].url.path, '/auth/v1/token');
      expect(http.requests[0].body, contains('"email":"ana@daypilot.test"'));
    });

    test('with no user in the response throws', () async {
      http.enqueue(jsonRow(const {}));

      expect(
        () => repo.login(email: 'ana@daypilot.test', password: 'wrong'),
        throwsA(isA<AuthException>()),
      );
    });
  });

  group('signUp', () {
    test('creates a new account and profile', () async {
      http.enqueue(jsonRow(_sessionBody('user-1', userMetadata: {'username': 'ana', 'name': 'Ana'})));
      http.enqueue(emptyRows());
      http.enqueue(noContent());
      http.enqueue(jsonRow({
        'id': 'user-1',
        'email': 'ana@daypilot.test',
        'name': 'Ana',
        'username': 'ana',
        'photo_url': null,
        'region': 'ES',
        'level': 1,
      }));

      final user = await repo.signUp(
        name: 'Ana',
        username: 'ana',
        email: 'ana@daypilot.test',
        password: 'password123',
        region: 'ES',
      );

      expect(user!.username, 'ana');
      expect(http.requests[2].body, contains('"username_lower":"ana"'));
    });

    test('with no session in the response requires email confirmation', () async {
      http.enqueue(jsonRow(const {}));

      expect(
        () => repo.signUp(name: 'Ana', username: 'ana', email: 'ana@daypilot.test', password: 'password123'),
        throwsA(isA<EmailConfirmationRequiredError>()),
      );
    });

    test('for an already-registered email whose retry-login succeeds still reports the conflict', () async {
      http.enqueue(postgrestError(message: 'User already registered', statusCode: 400));
      http.enqueue(jsonRow(_sessionBody('user-1')));
      http.enqueue(jsonRows([
        {'id': 'user-1'},
      ]));

      expect(
        () => repo.signUp(name: 'Ana', username: 'ana', email: 'ana@daypilot.test', password: 'password123'),
        throwsA(isA<AuthException>()),
      );
    });

    test('for an already-registered email with the wrong password reports the conflict', () async {
      http.enqueue(postgrestError(message: 'User already registered', statusCode: 400));
      http.enqueue(postgrestError(message: 'Invalid login credentials', statusCode: 400));

      expect(
        () => repo.signUp(name: 'Ana', username: 'ana', email: 'ana@daypilot.test', password: 'wrong'),
        throwsA(isA<AuthException>()),
      );
    });

    test('rethrows an unrelated signup failure as-is', () async {
      http.enqueue(postgrestError(message: 'Signups not allowed for this instance', statusCode: 400));

      expect(
        () => repo.signUp(name: 'Ana', username: 'ana', email: 'ana@daypilot.test', password: 'password123'),
        throwsA(isA<AuthException>()),
      );
    });
  });

  group('logout', () {
    test('with an active session also revokes it server-side', () async {
      await http.signIn('user-1');
      http.enqueue(noContent());

      await repo.logout();

      expect(http.requests, hasLength(1));
      expect(http.requests[0].url.path, '/auth/v1/logout');
    });

    test('with no active session makes no request', () async {
      await repo.logout();

      expect(http.requests, isEmpty);
    });
  });

  test('sendPasswordResetEmail posts to the recovery endpoint', () async {
    http.enqueue(noContent());

    await repo.sendPasswordResetEmail('ana@daypilot.test');

    expect(http.requests[0].url.path, '/auth/v1/recover');
    expect(http.requests[0].body, contains('"email":"ana@daypilot.test"'));
  });
}

File: app/src/test/java/com/example/daypilot_test_desing/data/supabase/SupabaseAuthRepositoryTest.kt
package com.example.daypilot_test_desing.data.supabase

import com.example.daypilot_test_desing.core.data.repository.RegisterOutcome
import com.example.daypilot_test_desing.support.SharedFakeSupabaseClient
import io.ktor.client.engine.mock.MockRequestHandleScope
import io.ktor.client.engine.mock.respond
import io.ktor.client.engine.mock.toByteArray
import io.ktor.client.request.HttpRequestData
import io.ktor.client.request.HttpResponseData
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpMethod
import io.ktor.http.HttpStatusCode
import io.ktor.http.headersOf
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Before
import org.junit.Test

private val jsonHeaders = headersOf(HttpHeaders.ContentType, "application/json")

private fun sessionJson(userId: String = "u1", extraUserFields: String = "") = """
    {"access_token":"tok","refresh_token":"rtok","expires_in":3600,"token_type":"bearer",
     "user":{"aud":"authenticated","id":"$userId"$extraUserFields}}
""".trimIndent()

private fun pendingConfirmationUserJson(userId: String = "u1") =
    """{"aud":"authenticated","id":"$userId"}"""

private fun MockRequestHandleScope.alreadyRegisteredErrorResponse() =
    respond(
        """{"msg":"User already registered"}""",
        status = HttpStatusCode.UnprocessableEntity,
        headers = jsonHeaders
    )

private fun MockRequestHandleScope.wrongPasswordErrorResponse() =
    respond(
        """{"msg":"Invalid login credentials"}""",
        status = HttpStatusCode.BadRequest,
        headers = jsonHeaders
    )

class SupabaseAuthRepositoryTest {

    companion object {
        private val shared = SharedFakeSupabaseClient()
    }

    private fun fakeSupabaseClient(
        handler: suspend MockRequestHandleScope.(HttpRequestData) -> HttpResponseData
    ) = shared.use(handler)

    @Before
    fun setUp() = runBlocking {
        shared.resetAuth()
    }

    @Test
    fun `login signs in and creates a missing profile row`() = runTest {
        var insertedBody: String? = null
        val client = fakeSupabaseClient { request ->
            when {
                request.url.encodedPath.endsWith("/token") -> respond(sessionJson(), headers = jsonHeaders)
                request.url.encodedPath.endsWith("/users") && request.method == HttpMethod.Get ->
                    respond("[]", headers = jsonHeaders)
                request.url.encodedPath.endsWith("/users") && request.method == HttpMethod.Post -> {
                    insertedBody = String(request.body.toByteArray())
                    respond("[]", headers = jsonHeaders)
                }
                else -> respond("[]", headers = jsonHeaders)
            }
        }
        val repo = SupabaseAuthRepository(client)

        repo.login("a@x.com", "password123")

        assertTrue(insertedBody!!.contains("\"id\":\"u1\""))
    }

    @Test
    fun `login skips profile creation when the row already exists`() = runTest {
        val client = fakeSupabaseClient { request ->
            when {
                request.url.encodedPath.endsWith("/token") -> respond(sessionJson(), headers = jsonHeaders)
                request.url.encodedPath.endsWith("/users") && request.method == HttpMethod.Get ->
                    respond("""[{"id":"u1"}]""", headers = jsonHeaders)
                request.url.encodedPath.endsWith("/users") && request.method == HttpMethod.Post ->
                    error("should not insert when a profile row already exists")
                else -> respond("[]", headers = jsonHeaders)
            }
        }
        val repo = SupabaseAuthRepository(client)

        repo.login("a@x.com", "password123")
    }

    @Test
    fun `register creates the account and profile row on a fresh signup`() = runTest {
        var insertedBody: String? = null
        val client = fakeSupabaseClient { request ->
            when {
                request.url.encodedPath.endsWith("/signup") -> respond(sessionJson(), headers = jsonHeaders)
                request.url.encodedPath.endsWith("/users") && request.method == HttpMethod.Get ->
                    respond("[]", headers = jsonHeaders)
                request.url.encodedPath.endsWith("/users") && request.method == HttpMethod.Post -> {
                    insertedBody = String(request.body.toByteArray())
                    respond("[]", headers = jsonHeaders)
                }
                else -> respond("[]", headers = jsonHeaders)
            }
        }
        val repo = SupabaseAuthRepository(client)

        val outcome = repo.register("Ana", "ana", "a@x.com", "password123", "europe_madrid")

        assertEquals(RegisterOutcome.Success, outcome)
        assertTrue(insertedBody!!.contains("\"username_lower\":\"ana\""))
    }

    @Test
    fun `register returns pending confirmation when signup yields no session`() = runTest {
        val client = fakeSupabaseClient { request ->
            when {
                request.url.encodedPath.endsWith("/signup") ->
                    respond(pendingConfirmationUserJson(), headers = jsonHeaders)
                else -> respond("[]", headers = jsonHeaders)
            }
        }
        val repo = SupabaseAuthRepository(client)

        val outcome = repo.register("Ana", "ana", "a@x.com", "password123", "europe_madrid")

        assertEquals(RegisterOutcome.PendingEmailConfirmation, outcome)
    }

    @Test
    fun `register returns already exists when a profile row is already present`() = runTest {
        val client = fakeSupabaseClient { request ->
            when {
                request.url.encodedPath.endsWith("/signup") -> respond(sessionJson(), headers = jsonHeaders)
                request.url.encodedPath.endsWith("/users") && request.method == HttpMethod.Get ->
                    respond("""[{"id":"u1"}]""", headers = jsonHeaders)
                request.url.encodedPath.endsWith("/users") && request.method == HttpMethod.Post ->
                    error("should not insert when a profile row already exists")
                else -> respond("[]", headers = jsonHeaders)
            }
        }
        val repo = SupabaseAuthRepository(client)

        val outcome = repo.register("Ana", "ana", "a@x.com", "password123", "europe_madrid")

        assertEquals(RegisterOutcome.AlreadyExists, outcome)
    }

    @Test
    fun `register falls back to sign-in on an already-registered account and finishes the profile`() = runTest {
        var insertedBody: String? = null
        val client = fakeSupabaseClient { request ->
            when {
                request.url.encodedPath.endsWith("/signup") -> alreadyRegisteredErrorResponse()
                request.url.encodedPath.endsWith("/token") -> respond(sessionJson(), headers = jsonHeaders)
                request.url.encodedPath.endsWith("/users") && request.method == HttpMethod.Get ->
                    respond("[]", headers = jsonHeaders)
                request.url.encodedPath.endsWith("/users") && request.method == HttpMethod.Post -> {
                    insertedBody = String(request.body.toByteArray())
                    respond("[]", headers = jsonHeaders)
                }
                else -> respond("[]", headers = jsonHeaders)
            }
        }
        val repo = SupabaseAuthRepository(client)

        val outcome = repo.register("Ana", "ana", "a@x.com", "password123", "europe_madrid")

        assertEquals(RegisterOutcome.Success, outcome)
        assertTrue(insertedBody!!.contains("\"id\":\"u1\""))
    }

    @Test
    fun `register falls back to sign-in on an already-registered account with an existing profile`() = runTest {
        val client = fakeSupabaseClient { request ->
            when {
                request.url.encodedPath.endsWith("/signup") -> alreadyRegisteredErrorResponse()
                request.url.encodedPath.endsWith("/token") -> respond(sessionJson(), headers = jsonHeaders)
                request.url.encodedPath.endsWith("/users") && request.method == HttpMethod.Get ->
                    respond("""[{"id":"u1"}]""", headers = jsonHeaders)
                request.url.encodedPath.endsWith("/users") && request.method == HttpMethod.Post ->
                    error("should not insert when a profile row already exists")
                else -> respond("[]", headers = jsonHeaders)
            }
        }
        val repo = SupabaseAuthRepository(client)

        val outcome = repo.register("Ana", "ana", "a@x.com", "password123", "europe_madrid")

        assertEquals(RegisterOutcome.AlreadyExists, outcome)
    }

    @Test
    fun `register returns already exists when the sign-in fallback fails with the wrong password`() = runTest {
        val client = fakeSupabaseClient { request ->
            when {
                request.url.encodedPath.endsWith("/signup") -> alreadyRegisteredErrorResponse()
                request.url.encodedPath.endsWith("/token") -> wrongPasswordErrorResponse()
                else -> respond("[]", headers = jsonHeaders)
            }
        }
        val repo = SupabaseAuthRepository(client)

        val outcome = repo.register("Ana", "ana", "a@x.com", "password123", "europe_madrid")

        assertEquals(RegisterOutcome.AlreadyExists, outcome)
    }

    @Test
    fun `register rethrows signup failures unrelated to an existing account`() = runTest {
        val client = fakeSupabaseClient { request ->
            when {
                request.url.encodedPath.endsWith("/signup") ->
                    respond(
                        """{"msg":"Signups not allowed for this instance"}""",
                        status = HttpStatusCode.BadRequest,
                        headers = jsonHeaders
                    )
                else -> respond("[]", headers = jsonHeaders)
            }
        }
        val repo = SupabaseAuthRepository(client)

        try {
            repo.register("Ana", "ana", "a@x.com", "password123", "europe_madrid")
            fail("expected an exception to propagate")
        } catch (e: Exception) {
            assertTrue(e.message?.contains("Signups not allowed", ignoreCase = true) == true)
        }
    }

    @Test
    fun `sendResetEmail posts the recovery request`() = runTest {
        var requestedPath: String? = null
        val client = fakeSupabaseClient { request ->
            requestedPath = request.url.encodedPath
            respond("{}", headers = jsonHeaders)
        }
        val repo = SupabaseAuthRepository(client)

        repo.sendResetEmail("a@x.com")

        assertTrue(requestedPath!!.endsWith("/recover"))
    }

    @Test
    fun `ensureProfileExists falls back to metadata for name, username and region`() = runTest {
        var insertedBody: String? = null
        val client = fakeSupabaseClient { request ->
            when {
                request.url.encodedPath.endsWith("/token") -> respond(
                    sessionJson(
                        extraUserFields = ""","email":"ana@x.com","user_metadata":{"name":"Ana Garcia","username":"anag","region":"europe_madrid"}"""
                    ),
                    headers = jsonHeaders
                )
                request.url.encodedPath.endsWith("/users") && request.method == HttpMethod.Get ->
                    respond("[]", headers = jsonHeaders)
                request.url.encodedPath.endsWith("/users") && request.method == HttpMethod.Post -> {
                    insertedBody = String(request.body.toByteArray())
                    respond("[]", headers = jsonHeaders)
                }
                else -> respond("[]", headers = jsonHeaders)
            }
        }
        val repo = SupabaseAuthRepository(client)

        repo.login("ana@x.com", "password123")

        assertTrue(insertedBody!!.contains("\"name\":\"Ana Garcia\""))
        assertTrue(insertedBody!!.contains("\"username\":\"anag\""))
        assertTrue(insertedBody!!.contains("\"region\":\"europe_madrid\""))
    }

    @Test
    fun `ensureProfileExists derives username and name from the email when metadata is missing`() = runTest {
        var insertedBody: String? = null
        val client = fakeSupabaseClient { request ->
            when {
                request.url.encodedPath.endsWith("/token") -> respond(
                    sessionJson(extraUserFields = ""","email":"noMeta@x.com""""),
                    headers = jsonHeaders
                )
                request.url.encodedPath.endsWith("/users") && request.method == HttpMethod.Get ->
                    respond("[]", headers = jsonHeaders)
                request.url.encodedPath.endsWith("/users") && request.method == HttpMethod.Post -> {
                    insertedBody = String(request.body.toByteArray())
                    respond("[]", headers = jsonHeaders)
                }
                else -> respond("[]", headers = jsonHeaders)
            }
        }
        val repo = SupabaseAuthRepository(client)

        repo.login("noMeta@x.com", "password123")

        assertTrue(insertedBody!!.contains("\"username\":\"noMeta\""))
        assertTrue(insertedBody!!.contains("\"name\":\"noMeta\""))
        assertTrue(insertedBody!!.contains("\"region\":\"\""))
    }

    @Test
    fun `ensureProfileExists falls back to the user id when there is no email or username metadata`() = runTest {
        var insertedBody: String? = null
        val client = fakeSupabaseClient { request ->
            when {
                request.url.encodedPath.endsWith("/token") -> respond(sessionJson(userId = "u-no-email"), headers = jsonHeaders)
                request.url.encodedPath.endsWith("/users") && request.method == HttpMethod.Get ->
                    respond("[]", headers = jsonHeaders)
                request.url.encodedPath.endsWith("/users") && request.method == HttpMethod.Post -> {
                    insertedBody = String(request.body.toByteArray())
                    respond("[]", headers = jsonHeaders)
                }
                else -> respond("[]", headers = jsonHeaders)
            }
        }
        val repo = SupabaseAuthRepository(client)

        repo.login("whatever@x.com", "password123")

        assertTrue(insertedBody!!.contains("\"username\":\"u-no-email\""))
        assertTrue(insertedBody!!.contains("\"email\":\"\""))
    }
}
